File: server_poll.c
#include <stdio.h>
#include <time.h>
#include <string.h>
#include <errno.h>
#include <poll.h>
#include "socket_hdr.h"
#include "base64.h"

#define IP_ADDRESS "0.0.0.0"
#define PORT		20000
#define BUF_SIZE	64
#define MAX_POLL_SIZE 1024

int find_empty_postion_poll(const struct pollfd* arr, size_t len) {
    for(int i = 0; i < len;i++) {
        if(arr[i].fd == 0) {
            return i;
        }
    }
    return -1;
}

int main() {
    int svr_sock;                //套接字句柄
    int clifd;
    int result;
    int pos;
    char buf[BUF_SIZE];
    char* pB64;
    struct sockaddr_in addr;
    struct pollfd pollfds[MAX_POLL_SIZE];
    int timeout = 5000;
    int yes = 1;
    
    memset(pollfds, 0, sizeof(pollfds));

	if (init_socket() == -1) {
		printf("init socket error");
		return -1;
	}

	if ((svr_sock = socket(AF_INET, SOCK_STREAM, 0)) == 0){
		printf("create socket failed::%d\n", errno);
		return -1;
	}
    
    setsockopt(svr_sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));
    
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(PORT);
    if(bind(svr_sock, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
        printf("bind socket error\n");
        return -1;
    }
    
    if(listen(svr_sock, 5) == -1) {
        printf("listen socket error\n");
        return -1;
    }
    
    pollfds[0].fd = svr_sock;                //设置监控sockfd
    pollfds[0].events = POLLIN | POLLPRI;            //设置监控的事件
    for(;;){
        result = poll(pollfds, MAX_POLL_SIZE, timeout);
        switch(result){        //开始监控
            case -1:                    //函数调用出错
                printf("poll error \r\n");
                break;
            case 0:
                printf("time out \r\n");
                break;
            default:                    //得到数据返回
                for(int i = 0;i < MAX_POLL_SIZE;i++) {
                    if(i == 0 && pollfds[i].revents & POLLIN) {
                        clifd = accept(svr_sock, NULL, NULL);
                        printf("client accpet\n");
                        if((pos = find_empty_postion_poll(pollfds, MAX_POLL_SIZE)) == -1) {
                            printf("poll list full\n");
                        } else {
                            pollfds[pos].fd = clifd;
                            pollfds[pos].events = POLLIN|POLLPRI; 
                        }
                    } else if(i > 0 && pollfds[i].revents & POLLIN){
                        int n = recv(pollfds[i].fd, buf, 64, 0);
                        printf("n ====>> %d\n", n);
                        if(n == 0) {
                            close(pollfds[i].fd);
                            pollfds[i].fd = 0;
                             pollfds[i].revents = 0;
                       } else if(n == -1) {
                            close(pollfds[i].fd);
                            pollfds[i].fd = 0;
                             pollfds[i].revents = 0;
                        } else {
                            buf[n] = 0;
                            pB64 = base64_encode(buf);
				            printf("%s <> %s\n", pB64, buf);
                            sprintf(buf, "%s", base64_decode(pB64));
                            send(pollfds[i].fd, buf, strlen(buf), 0);
                        }
                    }
                }
                break;
        }
    }
}

File: server_poll.py
import select
import socket
import sys
import queue
from queue import Queue


def run_server():
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setblocking(False)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server_address = ('0.0.0.0', 20002)
    server.bind(server_address)
    server.listen(5)
    message_queues = {}

    # 常用的标识  代表你想检查的事件类型
    READ_ONLY = select.POLLIN | select.POLLPRI | select.POLLHUP | select.POLLERR
    READ_WRITE = READ_ONLY | select.POLLOUT

    TIMEOUT = 1000
    poller = select.poll() # 创建一个poll对象，该对象可以注册或注销文件描述符
    poller.register(server,READ_ONLY)
    fd_to_socket = { server.fileno(): server,}

    while True:
        events = poller.poll(TIMEOUT) # 如果timeout为None，将会阻塞，知道有事件发生
        if not events:
            print("poll超时无活动连接，重新轮询......")
            continue
        print("有" , len(events), "个新事件，开始处理......")

        for fd, flag in events:
            s = fd_to_socket[fd]
            if flag & (select.POLLIN | select.POLLPRI): # 有数据可以读取
                if s is server: # 表示有新的连接
                    # 可以读取数据
                    connection, client_address = s.accept()
                    print(sys.stderr, '新的连接来自:', client_address)
                    connection.setblocking(False)
                    fd_to_socket[connection.fileno()] = connection # 往fd字典中添加一个新的 文件描述符
                    poller.register(connection, READ_ONLY)
                    message_queues[connection] = Queue() # 为了防止等待客户端发来数据期间发生阻塞，分配一个队列用于保存数据
                else: # 表示客户端传来了消息
                    data = s.recv(1024)
                    if data: # 表明数据接受成功
                        print(sys.stderr, '接受数据 "%s" 来自 %s' % (data, s.getpeername()))
                        message_queues[s].put(data)
                        # 修改一个已经存在的fd，修改事件为写。这里表示服务器向客户端要发送数据
                        poller.modify(s, READ_WRITE)
                    else:
                        # 如果没有接受到数据，表示要断开连接
                        print(sys.stderr, '关闭', client_address, '并未读取到数据')
                        # 停止监听连接上的输入
                        poller.unregister(s)
                        s.close()

                        # 将此链接从队列中删除
                        del message_queues[s]

            elif flag & select.POLLHUP:
                print(sys.stderr, '关闭', client_address, '收到HUP后')
                poller.unregister(s)
                s.close()

            elif flag & select.POLLOUT:
                try:
                    next_msg = message_queues[s].get_nowait()
                except queue.Empty:
                    print(sys.stderr, '队列', s.getpeername(), '为空')
                    poller.modify(s, READ_ONLY)
                else:
                    print(sys.stderr, '发送 "%s" 到 %s' % (next_msg, s.getpeername()))
                    s.send(next_msg)

            elif flag & select.POLLERR:
                print(sys.stderr, '异常信息:', s.getpeername())
                poller.unregister(s)
                s.close()
                del message_queues[s]

if __name__ == "__main__":
    run_server()

